### ISSUE TITLE
asyncToGenerator not working on functions with return statements?

### ISSUE THREAD

**Issue by bolinfest** [6to5/6to5#625]
I am trying to use `{ "optional" : [ "asyncToGenerator"] }` because my target environment, Atom, supports generators and I don't want to add a dependency on the regnerator runtime.

I'm not sure if I'm using 6to5 v3.0.11 incorrectly or not, but my input:

```
async function foo() {
  return 42;
}
```

becomes:

```

var _asyncToGenerator = function(fn) {
    return function() {
        var gen = fn.apply(this, arguments);
        return new Promise(function(resolve, reject) {
            var callNext = step.bind(gen.next);
            var callThrow = step.bind(gen["throw"]);

            function step(arg) {
                try {
                    var info = this(arg);
                    var value = info.value;
                } catch (error) {
                    reject(error);
                    return;
                }
                if (info.done) {
                    resolve(value);
                } else {
                    Promise.resolve(value).then(callNext, callThrow);
                }
            }
            callNext();
        });
    };
};

var foo = _asyncToGenerator(function*() {
    return 42;
});
```

but when I call `foo()`, I get an error:

```
"TypeError: Method [Generator].prototype.next called on incompatible receiver undefined
  at next (native)
```

I believe that the result of `fn.apply(this, arguments)` needs to be checked to see whether it is the output of a generator. I don't know if you can leverage the work of http://taskjs.org/, but it seems like a good starting point.

Also, I thought regenerator had an option to target environments that support `yield`: is it possible to use that? Is that what is being used by `asyncToGenerator`?

**Comment by sebmck** [comment]
Possibly a c4fb5a5e403e419ab6b440e55cc18a5a0378f463 regression. The runtime taken from [here](https://github.com/benjamn/es7-async-await/blob/master/runtime.js) is used.

**Comment by bolinfest** [comment]
I believe this should also be valid:

```
async function foo() {
  return new Promise((resolve, reject) => resolve(42));
}
```

I'm 99% sure that regenerator does the right thing here.

**Comment by sebmck** [comment]
@bolinfest Not sure if that does what you think it does:

```javascript
foo().then(function (promise) {
  promise.then(function (42) {
    data === 42;
  });
});
```

**Comment by bolinfest** [comment]
Hmm, /cc @benjamn.

**Comment by bolinfest** [comment]
@sebmck It's my understanding/expectation that if a function is `async`, then it always declares a `Promise`. If the implementation of that function happens to include `return <promise>`, then I would expect the async function to return a Promise that resolves to the same value of `<promise>`.

**Comment by sebmck** [comment]
@bolinfest Returning a promise inside of an async function will just resolve and return the promise you just constructed, it doesn't delegate to it. To get the behaviour you're expecting you'd change your original function to:

```javascript
async function foo() {
  return await* new Promise((resolve, reject) => resolve(42));
}
```

**Comment by sebmck** [comment]
I figured the `Method [Generator].prototype.next called on incompatible receiver undefined` error out. It's probably a result of strict mode, I'll push out a fix.

**Comment by sebmck** [comment]
Fixed as of 3.0.14, thanks!

**Comment by bolinfest** [comment]
@sebmck Thanks for your speedy fixes, as always!
